test/extended/csrapprover/csrapprover.go

```diff
@@ -0,0 +1,163 @@
+	"k8s.io/client-go/kubernetes/scheme"
+	certclientv1beta1 "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+func init() {
+	certv1beta1.AddToScheme(scheme.Scheme)
+}
+
+var _ = g.Describe("don't pwn cluster via MCS and autoapprover:", func() {
+	// since login mutates the current kubeconfig we want to use NewCLI
+	// as that will give each one of our test runs a new config via SetupProject
+	oc := exutil.NewCLI("cluster-client-cert", exutil.KubeConfigPath())
+	defer g.GinkgoRecover()
+
+	g.It("deny pod's access to /config/master API endpoint", func() {
+		// the /config/master API port+endpoint is only visible from inside the cluster
+		// (-> we need to create a pod to try to reach it)  and contains the token
+		// of the node-bootstrapper, so no random pods should be able to see it
+		pod, err := exutil.NewPodExecutor(oc, "get-bootstrap-creds", "fedora:latest")
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// get the API server URL, mutate to internal API (use infra.Status.APIServerURLInternal) once API is bumped
+		infra, err := oc.AdminConfigClient().ConfigV1().Infrastructures().Get("cluster", metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		internalAPI, err := url.Parse(infra.Status.APIServerURL)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		internalAPI.Host = strings.Replace(internalAPI.Host, "api.", "api-int.", 1)
+		internalAPI.Host = strings.Replace(internalAPI.Host, ":6443", ":22623", 1)
+		internalAPI.Path = "/config/master"
+
+		// we should not be able to reach the endpoint
+		_, err = pod.Exec(fmt.Sprintf("curl -k %s", internalAPI.String()))
+		o.Expect(err).To(o.HaveOccurred())
+	})
+
+	g.It("node-approver SA token compromised, don't approve random CSRs with client auth", func() {
+		// we somehow were able to get the node-approver token, make sure we can't
+		// create node certs with client auth with it
+		priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		certRequestTemplate := &x509.CertificateRequest{
+			SignatureAlgorithm: x509.ECDSAWithSHA256,
+			Subject: pkix.Name{
+				CommonName:   "system:node:hacking-node.ec2.internal",
+				Organization: []string{"system:nodes"},
+			},
+			PublicKey: priv.PublicKey,
+		}
+
+		csrbytes, err := x509.CreateCertificateRequest(rand.Reader, certRequestTemplate, priv)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		csrSpec := certv1beta1.CertificateSigningRequestSpec{
+			Request: pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csrbytes}),
+			Usages: []certv1beta1.KeyUsage{
+				certv1beta1.UsageDigitalSignature,
+				certv1beta1.UsageKeyEncipherment,
+				certv1beta1.UsageClientAuth,
+			},
+			Groups: []string{
+				"system:serviceaccounts",
+				"system:serviceaccounts:openshift-machine-config-operator",
+				"system:authenticated",
+			},
+		}
+
+		kubeCSRString, err := getKubeCSRObjectString("some-csr", csrSpec)
```

Comment:
er you should not need this.  use the REST API directly to create the CSR

Reply:
Is there a way to pass the token to the REST API? I believe that once I use `oc.KubeClient()`, the `clientConfig` is read and all other options set by `Args()` are just thrown away

Reply:
Get the rest config, make anonymous copy, inject your token, create a CSR client